### ISSUE TITLE
Add op: remove-redundant-axioms

### ISSUE THREAD

**Issue by cmungall** [ontodev/robot#16]
For discussion, cc @dosumis

## Motivation

For various reasons, the release version of an ontology can end up with redundant axioms; e.g.

 1. `A SubClassOf B`
 2. `B SubClassOf C`
 3. `A SubClassOf C`

Here 3 is redundant. Generally this should be pruned from the public release. There may be situations where this is not desired: for example, 3 may be adorned with a useful axiom annotation

## How it is used

As part of a release process, after reasoning and after #7

Note on GO-centric corner case: should never be applied to editors ontology as it may cause "ontology autophagy" when used as part of a cached link approach.

## Extensions

The notion of redundancy can be extended to the existential graph too, but this is more subtle, and my be best considered out of scope for this op, and a new op developed. The name `perform-transitive-reduction` or similar should be reserved for this task and @fbastian's code adopted.

## Spec

For simplicity we will only consider the graph formed by SubClassOf axioms between named classes. Equivalence axioms are ignored for the purposes of determining redundancy. Thus in the following 1 will be retained

```
1. A SubClassOf B
2. A EquivalentTo B and R some Z
```

The user should perform #7 `relax` step first to ensure a complete SubClassOf axiom graph is present.  Motivation: if we use reasoning to determine redundancy it complicates structural and logical requirements (TBD).

If the ontology contains an axiom `A SubClassOf Z`, and there exist a chain of two or more axioms `A SubClassOf B`, `B SubClassOf ...` ... `... SubClassOf Z`, then `A SubClassOf Z` is considered redundant.

Alternate reasoner based way of finding redundancy that does not require pre-reasoning: If the ontology contains an asserted axiom `A SubClassOf Z` and two inferences, `A SubClassOf B`, 'B SubClassOf Z` (as determined by the owlapi reasoner api, i.e. proper SubClassOf), then A-Z is redundant.

All redundant axioms are found, and then removed

TODO: spec equivalence between named classes corner case

**Comment by fbastian** [comment]
Hi Chris, 

just wondering, why not using the OWLGraphManipulator class? OK, it might not be the most beautiful piece of code, and it uses a graph-walk approach, not a reasoner-based approach, but I think it respects pretty much all of your specs, or can be easily adapted to your needs.

It takes into account super object properties, indirect redundant relations, GCI axioms, and could easily targets SubClassOf axioms only. It is extensively unit tested, and has already been used a lot on Uberon.

**Comment by cmungall** [comment]
Your implementation is itself fine, it's just you were obliged to 
implement it within the behemoth that is owltools.

Some possibilities:

  1. add a dependency to owltools-core in robot
  2. do some refactoring to make an owltools-graph that is focused more 
purely on graph operations and add that as a dependency
  3. port Frederic's code to SciGraph (which will likely be our target 
for graphy-type operations)
  4. don't try and make robot do everything - consider graph operations 
out of scope and have the ontology pipeline call owltools for certain 
operations

Option 1 feels a little messy. You end up bringing in functionality that 
robot does better; for example, owltools-core inherits all kinds of 
awkward assumptions about OBO style IDs, whereas robot handles this in 
an extensible way with CURIEs. I guess this could simply be ignored.

Option 2 feels cleaner but may amount to busy work.

I like the idea of doing 3 regardless, but SciGraph feels like a big 
dependency for robot (but could bring additional benefits).

Option 4 takes us away from the goal of having everything necessary to 
make an ontology release or run an ontology CI check in one convenient 
package.

Opinions welcome, @hdietze @ccondit @fbastian @jamesaoverton

**Comment by jamesaoverton** [comment]
[SciGraph](https://github.com/SciGraph/SciGraph) looks interesting.

We've discussed #7, #8, and this issue a few times. I understand the goals in the abstract, but this isn't how I usually work with ontologies. I still feel like I don't grasp the details. In commit 8331259 I just implemented the redundant axiom removal that we do in OBI before publication.

It might be better for someone else to take the lead on this.

Looking at [OWLGraphManipulator.java](https://github.com/owlcollab/owltools/blob/2588368933912e42b9585306145a1f75f08978f6/OWLTools-Core/src/main/java/owltools/graph/OWLGraphManipulator.java) quickly, it looks like a lot of work to port it. There's a lot going on, and a lot of uses of OWLTools graph stuff. This does help me understand the details, though. Does this file cover the majority of the graph operations, or do some live elsewhere?

I know these operations are important, but I really don't know the best way to integrate them.

**Comment by fbastian** [comment]
I received a notification and I didn't realize it wasn't the owltools repository, sorry.
But it's true that OWLGraphManipulator also implements #7 and #8.

I would love to port it to something cleaner (solution 2 or 3), but I really don't have the time right now :/

**Comment by cmungall** [comment]
High priority for GO - @jamesaoverton can you add @hdietze to the team so he can assign himself to this

**Comment by cmungall** [comment]
Thinking ahead to extensions. There are different definitions of redundancy, for example, the redundancy addressed in the spec above is best described as `NamedClass-SubClass-graph-redundancy`. This would be in contrast to reducing the set of logical axioms to the minimal set that is entailment preserving.

This should be clear in the command usage, and the naming scheme we choose should be extensible.

Do we 

 1. use a new command for every definition
 2. use a single command with an optional flag to select the type, defaulting to SubClass graph

@jamesaoverton will advise as to which is most robotik.

Either way we want a sensible way of naming the possibilities. Formally, an axiom `A` is redundant with respect to some set of axioms `S` (where `S` is a subset of axioms in `S`) if `(S-A) |= A` (note there is no requirement that `A` is in `S`). So if we can come up with a sensible way of naming how `S` is derived from `O` this should guide our naming scheme.

The one of highest priority to GO (and simplest to implement) is the case where `S` comes from the set of SubClassOf axioms where neither the LHS or RHS is anonymous. aka `named-subclass-graph`, which is still a mouthful. Any ideas for shorthands @dosumis? I favour `is_a` since this already means subclass between NCs in obo format.

**Comment by jamesaoverton** [comment]
I've been try to use a single, simple verb for the name of each command, and then an option for selecting the type/mode/method, as required. So I prefer option 2.

**Comment by cmungall** [comment]
Important clarification.

The input ontology `O` has an *imports closure* `O*` (where `O` is a subset of `O*`)

 1. This operation will only remove axioms from `O`
 2. All axioms in `O*` will be used to calculate whether an axiom in `O` is redundant

**Comment by cmungall** [comment]
@jamesaoverton - I would like to turn this ticket into a spec for the operation. Possibly there would be a way to view this as some kind of extended `-h` option but this isn't such a high priority now, the main goal is to manage the specifications of the operations.

Options include:

 1. As javadoc in the java code for the operation
 2. As wiki pages in this repo
 3. As .md files in the master branch of this repo, in some doc/ directory

I am not very keen on (1). I prefer markdown, I prefer easy ways for others to suggest changes, and I think it's better to separate the spec from the implementation.

I marginally prefer (3) to (2). For either of these, the naming convention would be that the filename would precisely match the verb robot uses, possibly with some suffix or prefix, e.g. `remove-redundant-axioms-spec.md`.

**Comment by jamesaoverton** [comment]
I've started thinking about documentation files. The `docs/template.md` file is a start. It should have more examples, and they should all have similar sections.

Pandoc can convert Markdown to groff format for man pages, which might be helpful.

**Comment by cmungall** [comment]
We *could* write these in a `man` page style.. . but I was imagining something higher level and more in-depth and also decoupled from command line options.

**Comment by cmungall** [comment]
Some notes on the discussion above

The actual implementation spec is here:

https://github.com/ontodev/robot/blob/master/robot-core/src/main/java/org/obolibrary/robot/ReduceOperation.java

Note that this is done based purely on an OWL formalization of redundancy: no conversion to a separate graph model required. Note also that redundant existential axioms and redundant GCIs are handled.

The OWLGraphManipulator class is still stronger. Eg

```
1 A SubClassOf B
2 A SubClassOf part_of some Z
3 Z SubClassOf B
```

is not reduced by robot. OGM will remove 1.
